### ISSUE TITLE
Game result reporting on 'Exit to Windows' is still bugged

### ISSUE THREAD

**Issue by Askaholic** [FAForever/fa#3087]
Had some reports of unrated ladder games recently and I think the game result on premature exit is still a bit bugged. I am suspecting a concurrency issue. 

Here are some server logs for the game in question. We can see one player leaves immediately after being defeated, and incorrectly reports "defeat" for the winning player.

```
INFO     Apr 24  23:59:02 LadderGame.11574283            Added game connection GameConnection(Player(slonick81, 40118, (1290.49, 126.424), (870.0181639002014, 66.22869679105045)), Game(11574283,slonick81,maps/zerg_canyon_1v1.v0005.zip,0))
INFO     Apr 24  23:59:29 LadderGame.11574283            Added game connection GameConnection(Player(xzaguer, 140321, (931.36, 125.402), (852.15, 67.0526)), Game(11574283,slonick81,maps/zerg_canyon_1v1.v0005.zip,1))
INFO     Apr 24  23:59:35 GameConnection                 Launching game Game(11574283,slonick81,maps/zerg_canyon_1v1.v0005.zip,2) in state GameState.LOBBY
INFO     Apr 24  23:59:35 LadderGame.11574283            Game launched
INFO     Apr 25  00:18:09 LadderGame.11574283            40118 reported result for army 2: score 1
INFO     Apr 25  00:18:10 LadderGame.11574283            140321 reported result for army 2: score 1
INFO     Apr 25  00:18:10 LadderGame.11574283            40118 reported result for army 1: defeat -10
INFO     Apr 25  00:18:11 LadderGame.11574283            140321 reported result for army 1: defeat -10
INFO     Apr 25  00:18:17 LadderGame.11574283            40118 reported result for army 2: defeat -10
INFO     Apr 25  00:18:18 LadderGame.11574283            140321 reported result for army 2: victory 10
INFO     Apr 25  00:18:20 LadderGame.11574283            Removed game connection GameConnection(Player(slonick81, 40118, (1290.49, 126.424), (870.0181639002014, 66.22869679105045)), Game(11574283,slonick81,maps/zerg_canyon_1v1.v0005.zip,2))
INFO     Apr 25  00:18:21 LadderGame.11574283            Game finished normally
INFO     Apr 25  00:18:21 GameResults                    Multiple outcomes for game 11574283 army 2 resolved to GameOutcome.CONFLICTING. Reports are: defaultdict(<class 'set'>, {<GameOutcome.DEFEAT: 'defeat'>: {40118}, <GameOutcome.VICTORY: 'victory'>: {140321}})
INFO     Apr 25  00:18:21 LadderGame.11574283            Result for army 2, player: Player(xzaguer, 140321, (931.36, 125.402), (852.15, 67.0526)): score 1, outcome GameOutcome.CONFLICTING
INFO     Apr 25  00:18:21 LadderGame.11574283            Result for army 1, player: Player(slonick81, 40118, (1290.49, 126.424), (870.0181639002014, 66.22869679105045)): score 0, outcome GameOutcome.DEFEAT
INFO     Apr 25  00:18:21 LadderGame.11574283            Score for player Player(xzaguer, 140321, (931.36, 125.402), (852.15, 67.0526)): score 1, outcome GameOutcome.CONFLICTING
INFO     Apr 25  00:18:21 LadderGame.11574283            Score for player Player(slonick81, 40118, (1290.49, 126.424), (870.0181639002014, 66.22869679105045)): score 0, outcome GameOutcome.DEFEAT
INFO     Apr 25  00:18:21 GameResults                    Multiple outcomes for game 11574283 army 2 resolved to GameOutcome.CONFLICTING. Reports are: defaultdict(<class 'set'>, {<GameOutcome.DEFEAT: 'defeat'>: {40118}, <GameOutcome.VICTORY: 'victory'>: {140321}})
ERROR    Apr 25  00:18:21 LadderGame.11574283            Error during game end
INFO     Apr 25  00:19:05 LadderGame.11574283            Removed game connection GameConnection(Player(xzaguer, 140321, (931.36, 125.402), (852.15, 67.0526)), Game(11574283,slonick81,maps/zerg_canyon_1v1.v0005.zip,2))
```

There are a few situations in which the error can occurr:
1. Using the `exit to windows` button in the middle of an ongoing match
2. Using the `score` button before the game has ended
3. Using alt-f4? Need to test this one out, but it was a thought that I had

Basically any time the user exits the game before the sim has reached it's natural conclusion will cause the player to report a defeat for anyone left alive in the game at the time.

**Comment by Askaholic** [comment]
Apparently it may also be caused by simply going to the score screen before the "Game Over" message has appeared. So I guess it can happen whenever you exit the simulation before it has completed processing to the point where "Game Over" appears. 

Here are some server logs from a test performed by Blodir and ThomasHiatt. 

```
TRACE    Aug 05  23:38:08 LobbyConnection                << Blodir: {'command': 'GameResult', 'args': [1, 'defeat -10'], 'target': 'game'}
INFO     Aug 05  23:38:08 LadderGame.11623471            133636 reported result for army 1: defeat -10
TRACE    Aug 05  23:38:08 LobbyConnection                << ThomasHiatt: {'command': 'GameResult', 'args': [1, 'defeat -10'], 'target': 'game'}
INFO     Aug 05  23:38:08 LadderGame.11623471            115624 reported result for army 1: defeat -10
DEBUG    Aug 05  23:38:09 GameStatsService               Army result for Player(Blodir, 133636, (2508.76, 86.6876), (2406.5380718342954, 64.96835508282841)) => GameOutcome.DEFEAT 
TRACE    Aug 05  23:38:10 LobbyConnection                << Blodir: {'command': 'GameResult', 'args': [2, 'defeat -10'], 'target': 'game'}
INFO     Aug 05  23:38:10 LadderGame.11623471            133636 reported result for army 2: defeat -10
TRACE    Aug 05  23:38:13 LobbyConnection                << Blodir: {'command': 'GameState', 'args': ['Ended'], 'target': 'game'}
INFO     Aug 05  23:38:13 LadderGame.11623471            Removed game connection GameConnection(Player(Blodir, 133636, (2508.76, 86.6876), (2406.5380718342954, 64.96835508282841)), Game(11623471,Blodir,maps/canis3v3.v0001.zip,2))
DEBUG    Aug 05  23:38:13 GameConnection                 GameConnection(Player(Blodir, 133636, (2508.76, 86.6876), (2406.5380718342954, 64.96835508282841)), Game(11623471,Blodir,maps/canis3v3.v0001.zip,2)).abort()
TRACE    Aug 05  23:38:14 ServerContext                  ]]: {'command': 'game_info', 'visibility': 'public', 'password_protected': False, 'uid': 11623471, 'title': 'Blodir Vs ThomasHiatt', 'state': 'playing', 'featured_mod': 'ladder1v1', 'sim_mods': {}, 'mapname': 'canis3v3.v0001', 'map_file_path': 'maps/canis3v3.v0001.zip', 'host': 'Blodir', 'num_players': 2, 'max_players': 2, 'launched_at': 1596670632.602577, 'teams': {2: ['Blodir'], 3: ['ThomasHiatt']}}
TRACE    Aug 05  23:38:14 LobbyConnection                << ThomasHiatt: {'command': 'GameResult', 'args': [2, 'victory 10'], 'target': 'game'}
INFO     Aug 05  23:38:14 LadderGame.11623471            115624 reported result for army 2: victory 10
```

I will try to keep looking to see if I can find a foolproof way to prevent these bad results from being reported, but maybe this additional info is enough that @speed2CZ, @shalkya or @Strogoo has some idea of how to fix it as well. It sounds like Blodir and Thomas are able to reproduce it pretty reliably.

**Comment by speed2CZ** [comment]
Game log to those games would be nice to know whats going on there.

**Comment by Askaholic** [comment]
TBH I don't think the game logs are very helpful (I have produced them myself when investigating this issue). They just tell you the same information which you can already gather from the server logs, which is that the person who leaves sends bogus results for the other players in the game. There isn't really enough logging of the related code to get a better idea of what's going on.

**Comment by shalkya** [comment]
maybe we can prevent people from going to score screen until game has ended when we are in 1v1 ?

**Comment by Askaholic** [comment]
I found that the sync table has an attribute called `RequestingExit` which I’m hoping can be a drop in for my previous fix that makes it work in all situations. Gonna try and get some people to test it today.

**Comment by Askaholic** [comment]
Tested it, and `RequestingExit` is not set at the time that the results are reported.

**Comment by K-ETFreeman** [comment]
Dear askaholic
i just beg you to fix that in any way

Already few months i and all my friends is suffering from not-rated ladder
just another few not-rated wins in a row will make someone suicide maybe
do you need some data? check my last 2 ladder not rated wins, for example. As i remember, i pressed score button fast in https://replay.faforever.com/12680090 (сuz forgot about that bug) [NOT RATED WIN]
and waited to the end (pausing) of the game in https://replay.faforever.com/12695356 , another [NOT RATED WIN] (waiting hasn't help me or my opponent abused this bug)
i am so depressed that i can't play ladder now, while i can only lose rating or gain NOT RATED wins.

Heard that ThomasHiatt offered a solution. May be use it, even temporary?
This hell is longs for few months already, can you save peoples' countless brain cells from dying from depression?
P L E A S E

**Comment by Askaholic** [comment]
@K-ETFreeman We deployed a temporary fix in the server code yesterday, hopefully games will be rated while I keep trying to figure out how to fix it properly in the game code. The issue with leaving early will still be present though, the winner always needs to wait until they get the victory message.

**Comment by K-ETFreeman** [comment]
Many, many thanks for that!

**Comment by Crotalus** [comment]
@Askaholic 

You could try comment out this line:
https://github.com/FAForever/fa/blob/bc4e4286697d5bfd9c049edbfb27f4b0f3b7966f/lua/ui/dialogs/eschandler.lua#L22

And see if it stops sending bogus defeat message when someone exits the game

**Comment by Askaholic** [comment]
Yea that’s what I was going to mess around with next, because I noticed that `SessionEndGame` is also called when you exit to the score screen. I also was looking through the documented engine functions and noticed `SessionIsActive` which I was hoping would let me detect if that function was called or not, since I’m afraid removing it completely may have some unintended side effects like the game freezes we had before.

**Comment by Askaholic** [comment]
I tried the suggestion from @Crotalus. It does stop the game from reporting bad results, HOWEVER, it also makes the other players in the game get the "timeout" window where they have to wait for a minute and a half in order to kick the disconnected player. I imagine the reason it was added in the first place was to get rid of this, unfortunately it also causes bad result reporting.

If anyone like @KionX or @kyx0r knows what exactly `SessionEndGame` does under the hood, that would be really helpful. As far as I can tell it must do something like:
- Notify other clients that we are disconnecting [prevents connection timeout]
- Terminate the local sim [causes bad results]

If we can figure out how exactly it is doing the disconnection step, then we might be able to separate those two things out, and avoid triggering the stop sim step which causes all alive players to be reported as dead to the server.

Alternatively, if we can figure out a way to detect when it's been called then we can just prevent the `UserSync` code from reporting bad results. However, I tested `SessionIsActive` and it unfortunately does not do what I hoped. It seems to be `true` even after calling `SessionEndGame`.

**Comment by KionX** [comment]
I can't reproduce the error.
result: defeat occurs only for exiting players.
How do I reproduce this?

Sending false defeats can be achieved intentionally using the UI mod.

**Comment by Askaholic** [comment]
The way I’ve been doing it is to ctrl-k and then click the “exit to windows” button immediately after the comm explosion. You need to have at least 1 other player in the game. When you look in your log file, you will see that your game reports “defeat” for every player that was alive when you closed your game.

**Comment by KionX** [comment]
It didn't work.
ctrl-k and then click the “exit to windows”
I tried without mods in Skirmish and LAN Multiplayer with two or more players,
ctrl-k different players, before and after the ACU explosion.
result: defeat occurs only for exiting players.

**Comment by Askaholic** [comment]
I'll make a video of how I do it tomorrow, because I can reproduce it 100% of the time.

**Comment by Askaholic** [comment]
So for whatever reason as soon as I started trying to record a video, I was unable to reproduce it with the ctrl-k method. But I was able to reproduce it by simply exiting to windows without ctrl-k ing.

Here you can see a game, I'm army 1 dragun (is actually sheikah on draguns account on the test server) is army 2. When I exit the game, FA doesn't report any results for me (army 1) only for dragun (army 2) even though dragun did not die in the game.
https://youtu.be/JC7l6H_c5kY

**Comment by KionX** [comment]
I repeated everything on the video in LAN. Can't reproduce.
Pay attention to "===== EXITING THE GAME=====". This text is not in the FAF.
Also, there is an abnormally high amount of "send Ptr".
My log from army 1:
```
debug: Current gametime: 00:00:30
info: send Ptr: 0CCA0002 
info: send Ptr: 0CCA0002 
debug: Loading module '\000/lua/simping.lua\000'
info: terminated 
info: Sending game result: 1 defeat -10
debug: Loading module '\000/lua/ui/game/gameresult.lua\000'
info: GAMERESULT : \000defeat -10
debug: Loading module '\000/lua/ui/dialogs/score.lua\000'
debug: Loading module '\000/lua/ui/dialogs/hotstats.lua\000'
debug: Loading module '\000/lua/ui/dialogs/myeffecthelpers.lua\000'
info: Sending game result: 2 victory 10
info: GAMERESULT : \000victory 10
debug: Loading module '\000/effects/entities/uefnukeeffectcontroller01/uefnukeeffectcontroller01_script.lua\000'
debug: Loading module '\000/lua/sim/damagearea.lua\000'
debug: Loading module '\000/effects/entities/debrismisc04/debrismisc04_script.lua\000'
debug: Loading module '\000/lua/genericdebris.lua\000'
debug: Loading module '\000/effects/entities/uefnukeeffect01/uefnukeeffect01_script.lua\000'
debug: Loading module '\000/effects/entities/uefnukeeffect02/uefnukeeffect02_script.lua\000'
debug: Loading module '\000/effects/entities/uefnukeshockwave01/uefnukeshockwave01_script.lua\000'
debug: Loading module '\000/effects/entities/uefnukeshockwave02/uefnukeshockwave02_script.lua\000'
debug: Loading module '\000/effects/entities/uefnukeeffect03/uefnukeeffect03_script.lua\000'
debug: Loading module '\000/effects/entities/uefnukeflavorplume01/uefnukeflavorplume01_script.lua\000'
debug: Loading module '\000/lua/system/dkson.lua\000'
info: Sending stats: {"stats":[{"type":"Human","general":{"kills":{
```

**Comment by KionX** [comment]
It worked through the client.
```
debug: Current gametime: 00:00:30
info: ren_oblivion true
info: ren_ui false
info: send Ptr: E2FC0002 
info: recv Ptr: E2FC0002 
info: Callback packet received, exit sync is over 
debug: Loading module '\000/lua/simping.lua\000'
info: terminated 
info: Sending game result: 2 defeat -10
debug: Loading module '\000/lua/ui/game/gameresult.lua\000'
info: GAMERESULT : \000defeat -10
debug: Loading module '\000/lua/ui/dialogs/score.lua\000'
debug: Loading module '\000/lua/ui/dialogs/hotstats.lua\000'
debug: Loading module '\000/lua/ui/dialogs/myeffecthelpers.lua\000'
info: GPGNET: setting nat handler to 0x00000000
info: CNetTCPBuf::Read(): recv() failed: WSAEINTR
info: Run time: 0h02m18s
```

**Comment by kyx0r** [comment]
@KionX Try on pre-desync patch exe. Maybe there is some kind of corruption, as modified packet might be sent elsewhere? Then it might not be handled on that side, but I was pretty sure that those woudn't be affected. recv failed gives me weird wibes. You can easily disable the desync patch by removing the hook files from patch process, those will let the game use original code.

**Comment by kyx0r** [comment]
Those hooks would be hook_recvfrom.cpp hook_sendto.cpp hook_sim_dispatch.cpp hook_Gpg_Net_Entry.cpp hook_decode.cpp . Let me know if that makes difference

**Comment by KionX** [comment]
It doesn't seem to have anything to do with desync patch.
When exiting the game, the aibrain.lua->AbandonedByPlayer call is triggered only for the second player.

**Comment by KionX** [comment]
Exiting the game in LAN works correctly:
First, aibrain.lua->AbandonedByPlayer is called for the first player.
Then victory. lua->CheckVictory->line 110 for the second player.

When playing through the client, aibrain.lua->AbandonedByPlayer is called for the second player instead of the first.
AbandonedByPlayer is called by the engine.

**Comment by kyx0r** [comment]
Well, since the exact details for how game network code works remain unknown, the desync patch works by placing a 2 byte marker at the begging of the sendto/recv buffer, like so: https://github.com/kyx0r/FA-Binary-Patches/blob/0adc2062400f0f463d68979fc378a49f13b91581/sections/gpg_net.cpp#L210-L212

But corruption is not possible because from my research the first 32 bit address of input buf is used as 16bit, so lower 16 bits should be 0, and it it's not there are checks in place to make sure we only work with marker values and write only if those bits are 0. But if that sendto is being done many times, it can send modified packets to other places that don't handle them... But I did a symbol look up and recv in the engine appered only 2 times, the other place was code for lobby connection, so it should be unrelated. The reason sendptr is done many times is because we don't know how many clients there are, and we also don't know what the send requests are. Basically we send those markers everywhere until we get all responces, destinations are unknown to us, but they are known to the cpu. Well, in the essance of the hack lies a lack of knowledge of network system and just assumption that eventually the agreement will be reached.

**Comment by kyx0r** [comment]
Just do some tests with clean exe for now, to eliminate any chance of containmant. It may be unrelated after all. But since desync patch has a fundamental assumption in place, it may do weird things in theory.

**Comment by KionX** [comment]
Hmm. How can I use a clean exe in the client? He will redownload it.

**Comment by kyx0r** [comment]
Set the exe file to readonly

**Comment by KionX** [comment]
It works. Previously, the client crashed from this.

**Comment by KionX** [comment]
Clean exe works correctly.

**Comment by kyx0r** [comment]
I might be random, but that recv failed is definitely gives a wistle. I actually think there might be a better way to get those markers working in desync patch. We can stick them to the end of the send buffer instead, but the problem is memory, if the block was malloced to certain amount, we may not be able to add marker at the end, so to avoid bad accceses it will need a malloc + memcpy

**Comment by kyx0r** [comment]
One problem I found is that the receiving buffer can only read 0x200 bytes which is problematic here:
https://github.com/kyx0r/FA-Binary-Patches/blob/0adc2062400f0f463d68979fc378a49f13b91581/sections/gpg_net.cpp#L141

@KionX Would you be up to fix desync patch if it's actually problem cause here? You will need somehow find a way to transmit and receive extra data in that buffer.

**Comment by kyx0r** [comment]
the sendto may get buffers that are 16k in size if i remember correctly, so you may place the algorithm like so: check for len < 0x200 if that's the case extend the buffer and place markers, then when recv gets that buffer it will get return value of nbytes received in eax, check for markerks at eax offset, if there are decrement eax the size of markers. Anyway all depends you need to get in there with debugger and see what happened when recv gives and error.

**Comment by KionX** [comment]
I don't know much about desync patch and the network part of FAF and don't know when I can fix it.

**Comment by Sheikah45** [comment]
@Garanas can this be closed with https://github.com/FAForever/fa/pull/4016
